lib/LedgerSMB/PGTimestamp.pm

```diff
@@ -0,0 +1,223 @@
+
+__PACKAGE__->register(registry => 'default',
+                      types => ['timestamp', 'timestamptz']);
+
+sub _stringify {
+    my $self = shift;
+
+    # Stringify without the 'T' in the middle
+    #   which is the same way Pg stringifies
+    return $self->strftime('%F %T');
+}
+
+=head1 SUPPORTED FORMATS
+
+### TO DOCUMENT
+
+=cut
+
+
+=head1 CONSTRUCTOR SYNTAX
+
+LedgerSMB::PgDate->new({ date => DateTime->new(year => 2012, day => 31, month =>
+12)});
+
+Note the constructor here is private, and not intended to be called directly.
+
+Use from_db and from_input methods instead since these handle appropriately
+different formats and handle construction differently.
+
+=cut
+
+
+=head1 METHODS
+
+=over
+
+=item new
+
+Returns an empty timestamp object when the input is an empty string; otherwise
+defers object creation to the superclass.
+
+=cut
+
+sub new {
+    my $class = shift;
+    my @args = @_;
+
+    if (! @args) {
+        my $self = {};
+        bless $self, $class;
+
+        $self->is_date(0);
+        $self->is_time(0);
+
+        return $self;
+    }
+    return $class->SUPER::new(@args);
+}
+
+
+=item add_interval(string $interval, optional integer $n)
+
+This adds $n * $interval to the date, defaulting to 1 if $n is not supplied.
+
+=cut
+
+sub add_interval {
+    my ($self,$interval,$n) = @_;
+
+    my %delta_names = (
+        day => 'days',
+        week => 'weeks',
+        month => 'months',
+        quarter => 'months',
+        year => 'years',
+    );
+    my $delta_name = $delta_names{$interval};
+    #Validate asked interval
+    die "Bad interval: $interval" if not defined $delta_name;
+
+    $n //= 1;    # Default to 1
+    $n *= MONTHS_PER_QUARTER if $interval eq 'quarter'; # A quarter is 3 months
+
+    my $has_time = $self->is_time();
+    $self->add($delta_name => $n, end_of_month => 'preserve');
+    $self->is_time($has_time);  # Make sure that is_time sticks
+
+    return $self;
+}
+
+=item from_input($date_string, [$format])
+
+Parses this from an input string according to the user's dateformat,
+unless C<$format> has been specified to override it.
+
+=cut
+
+our $formats = {
+    'YYYY-MM-DD' => ['%F'],
+    'DD-MM-YYYY' => ['%d-%m-%Y', '%d-%m-%y'],
+    'DD.MM.YYYY' => ['%d.%m.%Y', '%d.%m.%y'],
+    'DD/MM/YYYY' => ['%d/%m/%Y', '%D'],
+    'MM-DD-YYYY' => ['%m-%d-%Y', '%m-%d-%y'],
+    'MM/DD/YYYY' => ['%m/%d/%Y', '%m/%d/%y'],
+    'MM.DD.YYYY' => ['%m.%d.%Y', '%m.%d.%y'],
+      'YYYYMMDD' => ['%Y%m%d'],
+        'YYMMDD' => ['%y%m%d'],
+      'DDMMYYYY' => ['%d%m%Y'],
+        'DDMMYY' => ['%d%m%y'],
+      'MMDDYYYY' => ['%m%d%Y'],
+        'MMDDYY' => ['%m%d%y'],
+     'DDmonYYYY' => ['%d%b%Y', '%d%b%y']
+};
+
+sub from_input{
+    my ($self, $input) = @_;
+
+    local $@ = undef;
+    return $input if eval {$input->isa(__PACKAGE__)} && $input->is_date;
+
+    return __PACKAGE__->new()
+        if ! $input; # matches undefined as well as ''
+
+    my $dt = eval { strptime('%F %T', $input) }
```

Comment:
The pod for this function says:

`Parses this from an input string according to the user's dateformat,
unless C<$format> has been specified to override it.`

But here we're parsing to a hard-coded format and don't seem to pay attention to `$format`.

Reply:
Yup. Thanks for pointing that out. I've removed the remark about the user's date format, replacing it by what we *do* support.